### ISSUE TITLE
client: context cancel causes ErrClusterUnavailable messages

### ISSUE THREAD

**Issue by mwitkow-io** [coreos/etcd#3208]
I tried using the `etcd/client` in my project and rely on a Context cancellation to interrupt a `Watcher`'s `Next` call.

There are two problems.

First, the vendoring `golang/context` means that the checks for Context errors inside client.go always return false, unless the user of `etcd/client` uses the same vendored context as Etcd:

In `(c *httpClusterClient) Do()`:
```go
		resp, body, err = hc.Do(ctx, action)
		if err != nil {
			cerr.Errors = append(cerr.Errors, err)
			if err == context.DeadlineExceeded || err == context.Canceled {
				fmt.Printf("etcd context error on Do: %v\n", err)
				return nil, nil, cerr
			}
			continue
		}
```
The above checks only work if the exact same symbols 

The second problem is that the `Error()` function of `ClusterError` always returns "etcd cluster is unavailable or misconfigured", regardless of what error caused it. Thus, discovering the underlying cause of failures is really really tricky.

**Comment by xiang90** [comment]
@mwitkow-io How do you think return a client.Canceled type error?

**Comment by mwitkow-io** [comment]
`context.Cancelled` is already defined, the biggest "wut" moment was that the `Error` representation doesn't say anything about the contained `Errors` when you print it.

I'm currently running into another weird issue that I can't figure out: 
I'm getting (when printing the content of `Errors`): 
`ClusterErr: [read tcp 127.0.0.1:2379: use of closed network connection]` right after I call `cancel()` on the context.

**Comment by xiang90** [comment]
@mwitkow-io 

We cannot return context.Cancelled since we have a vendor `context` pkg. The problem is `cancellation` is not a cluster error. We should return something else. 

> I'm currently running into another weird issue that I can't figure out: 

If you find out another issue, please open another pull request.

**Comment by mwitkow-io** [comment]
Ok, so there's the issue of `ClusterError.Error()` hiding the underlying cause.

And then there's the issue of `context` compatibility. How would `etcd`'s vendoring play with, for example gRPC? Let's assume I was making an etcd request from a gRPC context, and used the gRPC parent context in my etcd request. Ideally, the etcd request would be cancelled if the gRPC context got cancelled (incoming request timed out, or was torn down). But vendoring prevents that, because these are two independent types. 

Any ideas on how you guys could keep the benefits that vendoring gives you, while making this work?

**Comment by xiang90** [comment]
> Ok, so there's the issue of ClusterError.Error() hiding the underlying cause.

It does not hide the cause. It should return the correct generalized cause. Then you can assert the type and see the exact cause. Current problem is the error is incorrect. The cluster is OK, but the client returns an incorrect error. I do not think we should change the clusterError.

Ideally, the client library does not use vendor version of anything since it is a library. The client is not stable and still under active development. We need to try it out and get more feedbacks.

**Comment by mwitkow-io** [comment]
@xiang90: maybe just make the output of `Errors` to be what `Details()` now returns? It would make it way clearer. Currently it's really hard to even realise that we should be getting a `ClusterError` from any of the calls, so it's not obvious that we should make the assertion.

As for vendoring, I completely understand why you're doing it :) It just makes things slightly more cumbersome to use for the end user (see the gRPC example above). 

And I'm more than happy to provide feedback, I've been using the `etcd/client` instead of `go-etcd` for:
https://github.com/mwitkow-io/go-flagz/

**Comment by xiang90** [comment]
> maybe just make the output of Errors to be what Details() now returns?

We will try out that idea. Stay tuned.

**Comment by xiang90** [comment]
> Currently it's really hard to even realise that we should be getting a ClusterError from any of the calls, so it's not obvious that we should make the assertion.

I think this is a documentation issue. I think it is really important to provide a full list of structured errors. It could be important to users.

For the cancellation issue, see #3229 and #3216

**Comment by xiang90** [comment]
Fixed by #3262
